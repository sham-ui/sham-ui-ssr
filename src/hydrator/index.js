import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,
        new ServerSideHooks( storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
