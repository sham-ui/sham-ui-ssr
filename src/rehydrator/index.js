import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
