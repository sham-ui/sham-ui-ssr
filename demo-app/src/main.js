import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

