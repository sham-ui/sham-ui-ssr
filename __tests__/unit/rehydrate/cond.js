import { compile, compileAsSFC } from 'sham-ui-test-helpers';
import { ssrAndRehydrate } from '../../../src/testing';

beforeEach( () => {
    window.LinkTo = compile`
        <a href={{url}}>
            {% defblock %}
        </a>
    `;

    window.DisplayContent = compile`
        {% if condition %}
            {% defblock %}
        {% endif %}
    `;

    window.TextContent = compile`
        <span>
            {% defblock %}
        </span>
    `;
} );

afterEach( () => {
    delete window.LinkTo;
    delete window.DisplayContent;
    delete window.TextContent;
} );

it( 'should work with {% default %}', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <LinkTo>
                    {% default %}
                        Text for content
                    {% end default %}
                </LinkTo>
            </div>
        `,
        {}
    );
    expect( meta.html ).toBe(
        '<div><a> Text for content <!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with two named blocks', async() => {
    expect.assertions( 2 );
    window.CustomPanel = compile`
        <div>
            <div class="title">
                {% defblock title %}
            </div>
            <div class="content">
                {% defblock %}
            </div>
        </div>
    `;
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <CustomPanel>
                    {% title %}
                        Text for title
                    {% end title %}

                    {% default %}
                        Text for content
                    {% end default %}
                </CustomPanel>
            </div>
        `,
        {}
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><div><div class="title"> Text for title <!--0--></div><div class="content"> Text for content <!--1--></div></div></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.CustomPanel;
} );

it( 'should work with component arguments', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <LinkTo url={{url}}>
                    {% default %}
                        Text for {{url}}
                    {% end default %}
                </LinkTo>
            </div>
        `,
        {
            url: 'http://example.com'
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><a href="http://example.com"> Text for <!--0-->http://example.com<!--1--> <!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        url: 'http://foo.example.com'
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        '<div><a href="http://foo.example.com"> Text for <!--0-->http://foo.example.com<!--1--> <!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with component default block', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <div>
                <LinkTo url={{url}}>
                    Text for {{url}}
                </LinkTo>
            </div>
        `,
        {
            url: 'http://example.com'
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div><a href="http://example.com"> Text for <!--0-->http://example.com<!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        url: 'http://foo.example.com'
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        '<div><a href="http://foo.example.com"> Text for <!--0-->http://foo.example.com<!--0--></a></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should remove block in if', async() => {
    expect.assertions( 6 );
    window.VisibleBlock = compile`
        {% if visible %}
            <div class="content">
                {% defblock %}
            </div>
        {% endif %}
    `;
    const meta = await ssrAndRehydrate(
        compile`
            <VisibleBlock visible={{visible}}>
                Text content for {{data}}
            </VisibleBlock>
        `,
        {
            visible: true,
            data: 'foo'
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<div class="content"> Text content for <!--0-->foo<!--0--></div><!--0--><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        visible: false,
        data: 'foz'
    } );
    expect( meta.ctx.container.innerHTML ).toBe( '<!--0--><!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        visible: true,
        data: 'foo'
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        '<div class="content"> Text content for foo<!--0--></div><!--0--><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.VisibleBlock;
} );

it( 'should work with two nested if', async() => {
    expect.assertions( 6 );
    window.BigRedButton = compile`
        {% if big %}
            {% if red %}
                <button class="big red">This button big={{big}}, red={{red}}{% defblock %}</button>
            {% endif %}
        {% endif %}
    `;
    const meta = await ssrAndRehydrate(
        compile`
            <BigRedButton big={{big}} red={{red}}>
                big && red
            </BigRedButton>
        `,
        {
            big: false,
            red: false
        }
    );
    expect( meta.html ).toBe( '<!--0--><!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        big: true
    } );
    expect( meta.ctx.container.innerHTML ).toBe( '<!--0--><!--0--><!--0-->' );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        red: true
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        '<button class="big red">This button big=true, red=true big &amp;&amp; red <!--0--></button><!--0--><!--0--><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.BigRedButton;
} );

it( 'should work with defblock nested in useblock', async() => {
    expect.assertions( 5 );
    window.LoadedContainer = compile`
        {% if loaded %}
            {% defblock %}
        {% endif %}
    `;
    window.LoadedVisibleContainer = compile`
        <LoadedContainer loaded={{loaded}}>
            {% if visible %}
                {% defblock %}
            {% endif %}
        </LoadedContainer>
    `;
    window.RedLoadedVisibleContainer = compile`
        <LoadedVisibleContainer loaded={{loaded}} visible={{visible}}>
            {% if red %}
                {% defblock %}
            {% endif %}
        </LoadedVisibleContainer>
    `;
    const meta = await ssrAndRehydrate(
        compile`
            <RedLoadedVisibleContainer loaded={{loaded}} visible={{visible}} red={{red}}>
                red && loaded & visible
            </RedLoadedVisibleContainer>
        `,
        {
            red: false,
            loaded: false,
            visible: false
        }
    );
    expect( meta.html ).toBe(
        // eslint-disable-next-line max-len
        '<!--0--><!--0--><!--0--><!--0-->'
    );

    meta.component.update( {
        loaded: true
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        '<!--0--><!--0--><!--0--><!--0--><!--0--><!--0-->'
    );

    meta.component.update( {
        visible: true
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        '<!--0--><!--0--><!--0--><!--0--><!--0--><!--0--><!--0--><!--0-->'
    );
    meta.component.update( {
        red: true
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        // eslint-disable-next-line max-len
        ' red &amp;&amp; loaded &amp; visible <!--0--><!--0--><!--0--><!--0--><!--0--><!--0--><!--0--><!--0--><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
    delete window.LoadedContainer;
    delete window.LoadedVisibleContainer;
    delete window.RedLoadedVisibleContainer;
} );

it( 'should work with for', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`
            <ul>
                {% for url of links %}
                    <li>
                        <LinkTo url={{url}}>Text for {{url}}</LinkTo>
                    </li>
                {% endfor %}
            </ul>
            <LinkTo url="http://example.com">Home</LinkTo>
        `,
        {
            links: [
                'http://foo.example.com',
                'http://bar.example.com',
                'http://baz.example.com'
            ]
        }
    );
    expect( meta.html ).toBe(
        '<ul>' +
        // eslint-disable-next-line max-len
        '<li><a href="http://foo.example.com">Text for <!--0-->http://foo.example.com<!--0--></a></li>' +
        // eslint-disable-next-line max-len
        '<li><a href="http://bar.example.com">Text for <!--0-->http://bar.example.com<!--0--></a></li>' +
        // eslint-disable-next-line max-len
        '<li><a href="http://baz.example.com">Text for <!--0-->http://baz.example.com<!--0--></a></li>' +
        '</ul>' +
        '<a href="http://example.com">Home<!--0--></a><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.update( {
        links: [
            'http://baz.example.com',
            'http://bar.example.com',
            'http://foo.example.com'
        ]
    } );
    expect( meta.ctx.container.innerHTML ).toBe(
        '<ul>' +
        // eslint-disable-next-line max-len
        '<li><a href="http://baz.example.com">Text for <!--0-->http://baz.example.com<!--0--></a></li>' +
        // eslint-disable-next-line max-len
        '<li><a href="http://bar.example.com">Text for <!--0-->http://bar.example.com<!--0--></a></li>' +
        // eslint-disable-next-line max-len
        '<li><a href="http://foo.example.com">Text for <!--0-->http://foo.example.com<!--0--></a></li>' +
        '</ul>' +
        '<a href="http://example.com">Home<!--0--></a><!--0-->'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work useblock if was update from block component', async() => {
    expect.assertions( 4 );
    const meta  = await ssrAndRehydrate(
        compile`
            <DisplayContent>
                Content
            </DisplayContent>
        `,
        {

        }
    );
    expect( meta.ctx.container.textContent.trim() ).toBe( '' );

    const displayContent = Array.from( meta.DI.resolve( 'sham-ui:store' ).byId.values() ).find(
        x => x instanceof window.DisplayContent
    );
    displayContent.update( { condition: true } );
    expect( meta.ctx.container.textContent.trim() ).toBe( 'Content' );

    displayContent.update( { condition: false } );
    expect( meta.ctx.container.textContent.trim() ).toBe( '' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should correct resolve owner', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compileAsSFC`
            <template>
                <TextContent>
                    <TextContent>
                        {{this._text()}}
                    </TextContent>
                </TextContent>
            </template>

            <script>
                export default Component( Template, function( options ) {
                    options( {
                        text: () => 'Text for content'
                    } );
                    this._text = () => this.options.text();
                } );
            </script>
        `
    );
    expect( meta.ctx.container.textContent.trim() ).toBe( 'Text for content' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


