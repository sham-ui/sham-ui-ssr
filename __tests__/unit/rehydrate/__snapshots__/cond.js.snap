// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correct resolve owner 1`] = `
Object {
  "Options": Object {
    "text": [Function],
  },
  "html": "
<span><span>Text for content<!--0--></span>
<!--0-->
<!--0--></span>
<!--0-->
",
}
`;

exports[`should remove block in if 1`] = `
Object {
  "Options": Object {
    "data": "foo",
    "visible": true,
  },
  "html": "
<div class=\\"content\\"> Text content for
  <!--0-->foo
  <!--0-->
</div>
<!--0-->
<!--0-->
",
}
`;

exports[`should remove block in if 2`] = `
Object {
  "Options": Object {
    "data": "foz",
    "visible": false,
  },
  "html": "
<!--0-->
<!--0-->
",
}
`;

exports[`should remove block in if 3`] = `
Object {
  "Options": Object {
    "data": "foo",
    "visible": true,
  },
  "html": "
<div class=\\"content\\"> Text content for foo
  <!--0-->
</div>
<!--0-->
<!--0-->
",
}
`;

exports[`should work useblock if was update from block component 1`] = `
Object {
  "Options": Object {},
  "html": "
<!--0-->
<!--0-->
",
}
`;

exports[`should work with {% block "default" %} 1`] = `
Object {
  "Options": Object {},
  "html": "
<div>
  <a> Text for content
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with component arguments 1`] = `
Object {
  "Options": Object {
    "url": "http://example.com",
  },
  "html": "
<div>
  <a href=\\"http://example.com\\"> Text for
    <!--0-->http://example.com
    <!--1-->
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with component arguments 2`] = `
Object {
  "Options": Object {
    "url": "http://foo.example.com",
  },
  "html": "
<div>
  <a href=\\"http://foo.example.com\\"> Text for
    <!--0-->http://foo.example.com
    <!--1-->
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with component default block 1`] = `
Object {
  "Options": Object {
    "url": "http://example.com",
  },
  "html": "
<div>
  <a href=\\"http://example.com\\"> Text for
    <!--0-->http://example.com
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with component default block 2`] = `
Object {
  "Options": Object {
    "url": "http://foo.example.com",
  },
  "html": "
<div>
  <a href=\\"http://foo.example.com\\"> Text for
    <!--0-->http://foo.example.com
    <!--0-->
  </a>
</div>
",
}
`;

exports[`should work with defblock nested in useblock 1`] = `
Object {
  "Options": Object {
    "loaded": true,
    "red": true,
    "visible": true,
  },
  "html": "
 red &amp;&amp; loaded &amp; visible
 <!--0-->
 <!--0-->
 <!--0-->
 <!--0-->
 <!--0-->
 <!--0-->
 <!--0-->
 <!--0-->
 <!--0-->
",
}
`;

exports[`should work with for 1`] = `
Object {
  "Options": Object {
    "links": Array [
      "http://foo.example.com",
      "http://bar.example.com",
      "http://baz.example.com",
    ],
  },
  "html": "
<ul>
  <li>
    <a href=\\"http://foo.example.com\\">Text for
      <!--0-->http://foo.example.com
      <!--0-->
    </a>
  </li>
  <li>
    <a href=\\"http://bar.example.com\\">Text for
      <!--0-->http://bar.example.com
      <!--0-->
    </a>
  </li>
  <li>
    <a href=\\"http://baz.example.com\\">Text for
      <!--0-->http://baz.example.com
      <!--0-->
    </a>
  </li>
</ul>
<a href=\\"http://example.com\\">Home
  <!--0-->
</a>
<!--0-->
",
}
`;

exports[`should work with for 2`] = `
Object {
  "Options": Object {
    "links": Array [
      "http://baz.example.com",
      "http://bar.example.com",
      "http://foo.example.com",
    ],
  },
  "html": "
<ul>
  <li>
    <a href=\\"http://baz.example.com\\">Text for
      <!--0-->http://baz.example.com
      <!--0-->
    </a>
  </li>
  <li>
    <a href=\\"http://bar.example.com\\">Text for
      <!--0-->http://bar.example.com
      <!--0-->
    </a>
  </li>
  <li>
    <a href=\\"http://foo.example.com\\">Text for
      <!--0-->http://foo.example.com
      <!--0-->
    </a>
  </li>
</ul>
<a href=\\"http://example.com\\">Home
  <!--0-->
</a>
<!--0-->
",
}
`;

exports[`should work with two named blocks 1`] = `
Object {
  "Options": Object {},
  "html": "
<div>
  <div>
    <div class=\\"title\\"> Text for title
      <!--0-->
    </div>
    <div class=\\"content\\"> Text for content
      <!--1-->
    </div>
  </div>
</div>
",
}
`;

exports[`should work with two nested if 1`] = `
Object {
  "Options": Object {
    "big": false,
    "red": false,
  },
  "html": "
<!--0-->
<!--0-->
",
}
`;

exports[`should work with two nested if 2`] = `
Object {
  "Options": Object {
    "big": true,
    "red": false,
  },
  "html": "
<!--0-->
<!--0-->
<!--0-->
",
}
`;

exports[`should work with two nested if 3`] = `
Object {
  "Options": Object {
    "big": true,
    "red": true,
  },
  "html": "
<button class=\\"big red\\">This button big=true, red=true big &amp;&amp; red
  <!--0-->
</button>
<!--0-->
<!--0-->
<!--0-->
",
}
`;
